pycqed/instrument_drivers/physical_instruments/NanotecSMI33.py

```diff
@@ -0,0 +1,768 @@
+                       'and value 255 selects the adaptive step mode.'))
+
+        self.add_parameter(
+            'travel_distance',
+            label='Travel Distance',
+            unit='Steps',
+            get_cmd='s',
+            set_cmd='s{}',
+            get_parser=int,
+            vals=Ints(min_value=-100000000,
+                      max_value=+100000000),
+            docstring=('Travel distance in steps'
+                       'Min value: -100000000'
+                       'Max value: +100000000'
+                       'Can only be positive when in relative positioning mode'
+                       'In absolute positioning mode, this sets the target'
+                       'position'))
+
+        self.connect_message()
+
+    def ask(self, cmd: str) -> str:
+        # Case for 'long' commands
+        if cmd[0] == ':' and len(cmd) > 2:
+            motor_cmd = self._start_character + self.controller_id + cmd
+            motor_response = super().ask(motor_cmd)
+            response = motor_response.lstrip('0').split(cmd)[1].lstrip('=')
+            return response
+        # Case for all other normal commands
+        else:
+            motor_cmd = self._start_character + self.controller_id + 'Z' + cmd
+            motor_response = super().ask(motor_cmd)
+            response = motor_response.lstrip('0').split(cmd)[1].lstrip()
+            return response
+
+
+    def get_idn(self):
+        # Make sure that the controller responds by placing it back into transmit
+        # mode
+        response = super().ask(self._start_character + self.controller_id + '|1')
+        info = self.ask('v')
+        info = info.split('_')
+        return {'vendor': 'Nanotec',
+                'model': info[0],
+                'serial': None,
+                'firmware': info[2]}
+
+    def load_record_from_eeprom(self, index: int) -> None:
+        """
+        Load a record from the EEPROM
+        Args
+            index: integer between 1 and 32
+        :return:
+        """
+        self.write('y' + str(index))
+
+    def save_record_to_eeprom(self, index: int) -> None:
+        """
+        Save the current record to the EEPROM
+        Args
+            index: integer between 1 and 32
+        :return:
+        """
+        self.write('>' + str(index))
+
+    def start_motor(self):
+        self.write('A')
+
+    def stop_motor(self, ramp='Quickstop'):
+        types = {
+            'Quickstop': 0,
+            'Brake': 1,
+        }
+        # TODO: add error checking to ramp type
+        self.write('S' + str(types[ramp]))
+
+    def write(self, cmd: str) -> None:
+        # Handle case of command that disables responding to commands
+        if cmd == '|0':
+            motor_cmd = self._start_character + self.controller_id + cmd
+            super().write(motor_cmd)
+        else:
+            motor_cmd = self._start_character + self.controller_id + cmd
+            response =  super().ask(motor_cmd)
```

Comment:
Is this intentional that `write` expects a response from the controller?

A minor comment about class structure, I think it would be nice to place the `write`  function just above or below the `ask` function.

Reply:
This is something that we could still discuss. This motor controller is not designed to follow anything like VISA instrument standards, so it always echos every write command (so you could check that it understood). I tried to just write to the device without reading the response, but I always had problems with subsequent reads returning the echo of the previous write. Clearing all the visa read and write buffers after each operation did not help. In the end, this was the only reasonably simple way that I found to solve this problem.

Reply:
Ok, I think this is a good solution if the controller just want to echo everything.